src/screen.cpp

```diff
@@ -734,7 +734,18 @@ static void s_update(screen_t *scr, const wcstring &left_prompt, const wcstring
     // Output the left prompt if it has changed.
     if (left_prompt != scr->actual_left_prompt) {
         s_move(scr, 0, 0);
-        s_write_str(scr, left_prompt.c_str());
+        size_t start = 0;
+        size_t prev_nl = 0;
+        size_t next_nl = wcstring::npos;
+        while ((next_nl = left_prompt.find('\n', start)) != wcstring::npos) {
+            s_write_str(scr, left_prompt.substr(prev_nl, next_nl - prev_nl).c_str());
+            if (clr_eol) {
+                s_write_mbs(scr, clr_eol);
+            }
+            prev_nl = next_nl;
+            start = prev_nl + 1;
+        }
+        s_write_str(scr, left_prompt.substr(prev_nl).c_str());
```

Comment:
We can avoid allocation in the common case by using `left_prompt.c_str() + prev_nl`.

We *could* also avoid allocations from the `substr` above, if we temporarily swap out newlines for null bytes. 

```cpp
        while ((next_nl = left_prompt.find('\n', start)) != wcstring::npos) {
	        left_prompt.at(next_nl) = L'\0';
            s_write_str(scr, left_prompt.c_str() + prev_nl);
	        left_prompt.at(next_nl) = L'\n';
	        ...
        }
```

However, I don't think swapping null bytes is common in the fish code base, so when in doubt I'd leave that part as is. What do others think?

Reply:
Done for the common case. Agreed that swapping '\n' with '\0' seems dubious. If we're going that far, we might as well make left_prompt a wcstring_list_t instead of a single wcstring. WDYT?